lib/internal/modules/esm/loader.js

```diff
@@ -240,6 +259,30 @@ class DefaultModuleLoader {
    * @returns {{ format: string, url: URL['href'] }}
    */
   resolve(originalSpecifier, parentURL, importAssertions) {
+    if (this.#delegate) {
+      return this.#delegate.resolve(originalSpecifier, parentURL, importAssertions);
+    }
+    defaultResolve ??= require('internal/modules/esm/resolve').defaultResolve;
+
+    const context = {
+      __proto__: null,
+      conditions: this.#defaultConditions,
+      importAssertions,
+      parentURL,
+    };
+
+    return defaultResolve(originalSpecifier, context);
+  }
+
+  /**
+   * Just like `resolve` except synchronous. This is here specifically to support
+   * `import.meta.resolve` which must happen synchronously.
+   */
+  resolveSync(originalSpecifier, parentURL, importAssertions) {
+    if (this.#delegate) {
+      return this.#delegate.resolveSync(originalSpecifier, parentURL, importAssertions);
+    }
```

Comment:
@izaakschroeder do you mind adding @JakobJingleheimer (or all the commenters on this PR) to your repo so we can push commits? We could help with some of the cleanup to get this landed sooner.

Reply:
@GeoffreyBooth @JakobJingleheimer I DRY'd this up. I preferred to avoid adding extra parameters (e.g. `forceSync`) since I think synchronous resolution is a pretty special case that requires special handling and anyone implementing the `Delegate` interface needs to take care that said method is implemented properly. Keeping the existing parameter set also more correctly mirrors the hook API which I feel is more intuitive.

My 2¢ anyway.

Reply:
Let’s please not forget the request to rename `this.#delegate` to something less generic. Maybe `this.#registeredHooks`?

Reply:
I'm not too picky. I don't think the object actually constitutes registered hooks (since, for example, `register()` as one of the methods that the delegate implements is not a hook) but if this is better then it's fine with me.

Reply:
Skimming through the code, it seems like the delegate is what was previously the customized module loader class that extended the default module loader class. So maybe we rename the class `DefaultModuleLoader` to just `ModuleLoader` and then `this.#delegate` becomes `this.#customizedModuleLoader`? Or `this.#customizations`?

Reply:
+1 for `this.#customizations` 😁